pkg/controller/controller.go

```diff
@@ -471,10 +463,21 @@ func (c *Controller) runAction(ctx context.Context, t *tomb.Tomb, as *crv1alpha1
 			var msg string
 			if err == nil {
 				c.updateActionSetRunningPhase(ctx, as, p.Name())
+				progressTrackCtx, doneProgressTrack := context.WithCancel(ctx)
+				defer doneProgressTrack()
+				go func() {
+					// progress update is computed on a best-effort basis.
+					// if it exits with error, we will just log it.
+					if err := progress.TrackActionsProgress(progressTrackCtx, c.crClient, as.GetName(), as.GetNamespace(), p); err != nil {
+						log.Error().WithError(err)
+					}
+				}()
 				output, err = p.Exec(ctx, *bp, action.Name, *tp)
+				doneProgressTrack()
```

Comment:
do we need to call it again even after calling it from `defer` ?

Reply:
Yes, the `doneProgressTrack()` needs to be called as soon as `Exec()` is done. The `defer()` gets called once all the phase Exec is done and action is completed (or failed). When the next phase execution starts during interaction in this function, the older phase's Progress() starts sending completion percent as 0. So before running next phase, the goroutine for current phase needs to be closed. Let me know if this makes sense or not.